### ISSUE TITLE
"Open with code" actions and PATH are not set for other accounts of a computer after install

### ISSUE THREAD

**Issue by sir-Higgins** [microsoft/vscode#25886]
- VSCode Version: 1.11.2
- OS Version: windows 10 1607

Steps to Reproduce:

When you install with an admin account that isn't your main account.
the "Open with code" action are applied on the account which install vscode.

![image](https://cloud.githubusercontent.com/assets/24956907/25699046/d2741d02-30c1-11e7-952f-c8cb0667ac46.png)

My main account is not the admin account that installed vscode
"Open with code" action to Windows Explorer file context menu,
"Open with code" action to Windows Explorer directory context menu,
Add code to PATH,
are not set to my account

Can you add a menu entry in vscode for setting the following actions for the current user that is running vscode afterward

"Open with code" action to Windows Explorer file context menu
"Open with code" action to Windows Explorer directory context menu
Add code to PATH

**Comment by ihalip** [comment]
I could look into this.

**Comment by Kas1894** [comment]
So with respect to the PATH variable issue and not being able to launch code from command line. It looks like the issue is due to how the setup.exe installs VSCode, like @ihalip said in his first example.

I have a similar setup to @sir-Higgins where my day to day account is non-admin. I went ahead and did a clean install of VS Code on my system (making sure to clear out my PATH variables on both admin and non-admin accounts). The results are show in the screenshots below. 

From left to right, the shells show the PATH variable contents for my non-admin, and admin accounts. The other window that's open shows the contents of the System Environment Variable. This was before VSCode was installed
![1](https://cloud.githubusercontent.com/assets/7507591/25774058/d6580244-3256-11e7-9c95-4455319c9445.PNG)

Checked the necessary boxes.
![2](https://cloud.githubusercontent.com/assets/7507591/25774060/d65a2416-3256-11e7-8396-75f303508837.PNG)

Here we see that an entry for `C:\Program Files (x86)\Microsoft VS Code\bin` has been added under the admin user (highlighted in red) but NOT for my non-admin user. 
![3](https://cloud.githubusercontent.com/assets/7507591/25774061/d65a4cfc-3256-11e7-9807-754d4fce9d75.PNG)

Obviously, this results in the inability to launch VSCode from the command line under the non-admin user.
![4](https://cloud.githubusercontent.com/assets/7507591/25774059/d6595234-3256-11e7-994d-b0df80509aac.PNG)

As a fix for this, I ended up adding `C:\Program Files (x86)\Microsoft VS Code\bin` to the System Environment PATH variable, which allowed me to launch VSCode from the command line as a non-admin user.

TL;DR: It looks to be caused by the installer inserting `C:\Program Files (x86)\Microsoft VS Code\bin` into the "wrong" users PATH.

**Comment by ihalip** [comment]
BTW, I'm testing on Windows 7.

Added this piece of code inside `code.iss`:
```
procedure CurStepChanged(CurStep: TSetupStep);
var
  ResultCode: Integer;
begin
  if CurStep = ssPostInstall then
  begin
    ExecAsOriginalUser('notepad.exe', '', '', SW_SHOW, ewWaitUntilTerminated, ResultCode);
  end;
end;
```

As expected, this doesn't work in the second use case I mentioned. If an account is not an Administrator, Windows doesn't even request admin credentials when trying to start setup.exe, it just shows a message box:
![image](https://cloud.githubusercontent.com/assets/1991437/25783119/21e321d6-335f-11e7-9869-f9ee7e5fa98d.png)

I can confirm that using `Run as a different user...` makes `ExecAsOriginalUser` actually run notepad.exe as the admin account and not the current user account. Furthermore, the last 'Run Code' page also starts VSCode as the admin account:
![image](https://cloud.githubusercontent.com/assets/1991437/25783082/c4bed612-335e-11e7-9cc6-1aeb321cc352.png)

I don't see an easy fix for this issue.

What I think could work is adding a couple of command line options to VSCode, i.e. `--add-to-path`, `--associate-with-files`, `--associate-with-dirs`, then adding an entry in the `[Run]` section in code.iss that would call code.exe with the appropriate command line options. There seems to be already support for a similar functionality for MacOS: https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/cli/electron-browser/cli.contribution.ts

Alternatively, a different script (batch? powershell?) could be created to do this work. Inno Setup code seems to have too many limitations.

**Comment by Kas1894** [comment]
I should've mentioned that I'm on Windows 10. 

Moving to a alternate installer with better(?) capabilities could be something to consider, though I have honestly no idea how much work that entails. Microsoft's own Orca installer (included with Windows SDK) is also freeware and an in-house option. 

However, based on what I [read here](https://stackoverflow.com/questions/25289056/how-to-set-a-global-environment-variable-from-inno-setup-installer), Inno Setup **should** have the ability to add `C:\Program Files (x86)\Microsoft VS Code\bin` to the system environment PATH variable, there by sidestepping the issue entirely. At that point, regardless of if an admin ran the installer, the "launch code from CLI" functionality would be available to all users. Unless that course of action is ill-advisable for reasons I don't know, it's probably the path of last resistance.

**Comment by sir-Higgins** [comment]
Maybe the problem is not to be solved in the installation process. The file/folder action and path can be set at the first run of vscode by a user after the installation.

When vscode is started for the first time he has to create the configuration files in the home directory at the same time he can ask the user if he want to set these settings.

This way this also work for new users on the computer. 
This would be useful for company computer that are on a domain.

**Comment by ihalip** [comment]
> Inno Setup should have the ability to [...]

It does, since the installer will always run as administrator. However, these settings are applied to HKEY_CURRENT_USER, which in these scenarios is not the the user which is logged in, but the user under which the installer runs.

**Comment by unional** [comment]
I would like to have these available to all users in the machine.
@sir-Higgins suggestion would be a good way to solve this.
Maybe adding an explicit command in [Ctrl+p] too?

**Comment by Kas1894** [comment]
I agree. Perhaps it can be added as part of the welcome screen that VS Code has? It could execute the following command in Windows, which would set the PATH properly for the current (non-admin) user: `setx PATH "%PATH%C:\Program Files\Microsoft VS Code\bin"`.

If it was going to be added as keyboard shortcut, I wouldn't do [ctrl+p] since that shortcut is generally used for invoking printing related things. Maybe something like ALT+P? But why would someone need to invoke the command more than once? You really only need to set the PATH variable once and then you're done.

**Comment by unional** [comment]
> I wouldn't do [ctrl+p] since

Sorry, I meant [ctrl+shift+p], in the command palette. No need to have a keyboard shortcut.

**Comment by AeliusSaionji** [comment]
Correct me if I'm wrong, but I think the discussion about "getting innosetup to know which user ran the install" is entirely a red herring.

The answer is as simple as using HKLM instead of HKCU in the install script.

In fact, is there any reason the script uses HKCU at all? Why would the installer only associate vscode with files under one user? That doesn't make sense.

I'll submit a PR.
